resources/type-system/inference.md

```diff
@@ -0,0 +1,830 @@
+
+## Local type inference
+
+When types are omitted on local variable declarations and function expressions,
+or when type arguments are omitted from literal expressions, constructor
+invocations, or generic function invocations, then local type inference is used
+to fill in the missing type information.  Local type inference is also used as
+part of top-level inference as described above, in order to infer types for
+initializer expressions.  In order to uniformly treat use of local type
+inference in top-level inference and in method body inference, it is defined
+with respect to a set of *available* variables as defined above.  Note however
+that top-level inference never depends on method body inference, and so method
+body inference can be performed as a subsequent step.  If this order of
+inference is followed, then method body inference should never fail due to a
+reference to an *unavailable* variable, since local variable declarations can
+always be traversed in an appropriate statically pre-determined order.
+
+### Types
+
+We define inference using types as defined in
+the
+[informal specification of subtyping](https://github.com/dart-lang/language/blob/master/resources/type-system/subtyping.md),
+with the same meta-variable conventions.  Specifically:
+
+The meta-variables `X`, `Y`, and `Z` range over type variables.
+
+The meta-variable `L` ranges over lists or sets of type variables.
+
+The meta-variables `T`, `S`, `U`, and `V` range over types.
+
+The meta-variable `C` ranges over classes.
+
+The meta-variable `B` ranges over types used as bounds for type variables.
+
+
+
+### Type schemas
+
+Local type inference uses a notion of `type schema`, which is a slight
+generalization of the normal Dart type syntax.  The grammar of Dart types is
+extended with an additional construct `?` which can appear anywhere that a type
+is expected.  The intent is that `?` represents a component of a type which has
+not yet been fixed by inference.  Type schemas cannot appear in programs or in
+final inferred types: they are purely part of the specification of the local
+inference process.  In this document, we sometimes refer to `?` as "the unknown
+type".
+
+It is an invariant that a type schema will never appear as the right hand
+component of a promoted type variable `X & T`.
+
+The meta-variables `P` and `Q` range over type schemas.
+
+### Variance
+
+We define the notion of the covariant, contravariance, and invariant occurrences
+of a type `T` in another type `S` inductively as follows.  Note that this
+definition of variance treats type aliases transparently: that is, the variance
+of a type which is used as an argument to a type alias is computed by first
+expanding the type alias (substituting actuals for formals) and then computing
+variance on the result.  This means that the only invariant positions in any
+type (given the current Dart type system) are in the bounds of generic function
+types.
```

Comment:
This is identical to the notion of variance ('occurs covariantly/contravariantly/invariantly') in the spec, except for the unfolding of type aliases (where a type parameter can be invariant because it is used in a covariant as well as a contravariant position, not just because it is used as a bound in a function type).

So how much does that matter?

Do we want to take this approach also in the spec? I suspect that any changes in this area will be so much of a breaking change that we won't do it, unless we have a really, really good reason. But could we then use the approach from the spec in this context?

The spec only uses this approach (of not unfolding type aliases when computing the variance of an occurrence) because i2b was designed to work in a way that can be considered to amount to provision of actual type arguments, and if there is no type argument which will produce a given result then i2b cannot give that result.

For example, with `typedef F<X> = X Function(X)`, i2b on `F` cannot yield `dynamic Function(Null)`. But i2b _will_ replace the same type variable by different types when the occurrences have different variance:

```dart
class C<X extends X Function(X)> {}

main() {
  C foo() => null;
  print(foo.runtimeType); // Prints '() => C<(Null) => dynamic>'.
}
```

I suspect that type inference can never produce a similar phenomenon (because it must yield a result which is a binding of a set of type variables to their inferred values, so it's immediate that each type variable has exactly one value).

So why is it useful (or necessary) to treat variance in this manner here?

Reply:
We can choose to treat typedefs transparently or opaquely.  That is, we can say that given `typedef F<T> = T Function(T)` that `F<X>` is in all places replaceable by `X Function(X)` (treat it transparently), or we can choose to allow it to behave differently (treat it opaquely).  For the purposes to which this is relevant in this document, treating them opaquely results in strictly worse results.  The greatest closure of a type with respect to `X` should be the type which is a supertype of all types which result from substituting a concrete type for `X`.  `Object Function(Null)` has that property for `F`, but `F<Object>` does not.  In general, invariant types force us compromise, either by not producing an upper bound of all instantiations, or by shifting to something like `Object` which loses more information.  This is fine if we need to, but here we don't need to.  I simply see no reason whatsoever to compute worse results here - why not just expand out the typedef and produce the correct result?